File: day:49/reverseLinkedList.js
/**
 * Definition for singly-linked list.
 * function ListNode(val, next) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.next = (next===undefined ? null : next)
 * }
 */
/**
 * @param {ListNode} head
 * @return {ListNode}
 */
var reverseList = function(head) {
    // base case
    if(head === null ||head.next === null) return head;
    
    let current = head;
    let next = current.next;
    
    while(next !== null) {
        let temp = next.next;
        next.next = current;
        current = next;
        next = temp;
    }
    head.next = null;
    head = current;
    
    return head;
};

File: day:49/reverseLinkedList.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def reverseList(self, head: ListNode) -> ListNode:
        # base case
        if(head == None or head.next == None): return head;
    
        current = head;
        next = current.next;
    
        while(next != None):
            temp = next.next;
            next.next = current;
            current = next;
            next = temp;
        head.next = None;
        head = current;
    
        return head;
